python/mxnet/gluon/data/text.py

```diff
@@ -0,0 +1,160 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# coding: utf-8
+# pylint: disable=
+"""Text datasets."""
+__all__ = ['WikiText2', 'WikiText103']
+
+import io
+import os
+import zipfile
+import shutil
+import numpy as np
+
+from . import dataset
+from ..utils import download, check_sha1
+from ...contrib import text
+from ... import nd
+
+
+class WikiText2(dataset._DownloadedDataset):
+    """WikiText-2 word-level dataset for language modeling, from Salesforce research.
+
+    From
+    https://einstein.ai/research/the-wikitext-long-term-dependency-language-modeling-dataset
+
+    License: Creative Commons Attribution-ShareAlike
+
+    Each sample is a vector of length equal to the specified sequence length.
+    At the end of each sentence, an end-of-sentence token '<eos>' is added.
+
+    Parameters
+    ----------
+    root : str, default '~/.mxnet/datasets/cifar10'
+        Path to temp folder for storing data.
+    segment : str, default 'train'
+        Dataset segment. Options are 'train', 'validation', 'test'.
+    indexer : :class:`~mxnet.contrib.text.indexer.TokenIndexer`, default None
```

Comment:
I wouldn't expose this to users.
1. Indexer is not the standard term for this.
2. This is contrib API and subject to change. Gluon Dataset should use a separate vocabulary API

Reply:
Thanks.
1. I think it's safe to say that indexer is a clear enough term to reflect what it does.
2. If I understand correctly, I believe the indexer class is intended to serve the same purpose as what you call 'vocabulary'

There is a reason this needs to be exposed. Suppose we have training dataset whose vocabulary is {a, b, c} plus unknown tokens, an test dataset has vocabulary {a, b, d}. As a standard practice, the token 'd' in the test dataset should be indexed as unknown. This means the indexing of test dataset depends on the index from training dataset.

Reply:
1. I don't think so. Do you have reference of it being used somewhere?
2. It is, but it is contrib API. If you want to use it directly then gluon.data.text need to be in gluon.contrib too.

We do need to expose something like this. But it can't be TokenIndexer.

Reply:
Do you have something else in mind?

Also, what should I provide here in place of TokenIndexer? If you could help me understand the reasoning for "it can't be TokenIndexer", I can help propose alternatives too.

Reply:
I think it's probably a safer bet to move the dataset to contrib first.